python/cudf_polars/cudf_polars/experimental/explain.py

```diff
@@ -0,0 +1,129 @@
+
+
+def explain_query(
+    q: pl.LazyFrame, engine: pl.GPUEngine, *, physical: bool = True
+) -> str:
+    """
+    Return a formatted string representation of the IR plan.
+
+    Parameters
+    ----------
+    q : pl.LazyFrame
+        The LazyFrame to explain.
+    engine : pl.GPUEngine
+        The configured GPU engine to use.
+    physical : bool, default True
+        If True, show the physical (lowered) plan.
+        If False, show the logical (pre-lowering) plan.
+
+    Returns
+    -------
+    str
+        A string representation of the IR plan.
+    """
+    config = ConfigOptions.from_polars_engine(engine)
+    ir = Translator(q._ldf.visit(), engine).translate_ir()
+
+    if physical:
+        lowered_ir, partition_info = lower_ir_graph(ir, config)
+        return _repr_ir_tree(lowered_ir, partition_info)
+    else:
+        return _repr_ir_tree(ir)
+
+
+def _repr_ir_tree(
+    ir: IR,
+    partition_info: MutableMapping[IR, PartitionInfo] | None = None,
+    *,
+    offset: str = "",
+) -> str:
+    header = _repr_ir(ir, offset=offset)
+    count = partition_info[ir].count if partition_info else None
+    if count is not None:
+        header = header.rstrip("\n") + f" [{count}]\n"
+
+    children_strs = [
+        _repr_ir_tree(child, partition_info, offset=offset + "  ")
+        for child in ir.children
+    ]
+
+    return header + "".join(
+        f"{line}{offset}  (repeated {count} times)\n"
+        if (count := sum(1 for _ in group)) > 1
+        else line
+        for line, group in groupby(children_strs)
+    )
+
+
+def _repr_schema(schema: tuple | None) -> str:
+    if schema is None:
+        return ""  # pragma: no cover; no test yet
+    names = tuple(schema)
+    if len(names) > 6:
+        names = names[:3] + ("...",) + names[-2:]
+    return f" {names}"
+
+
+def _repr_header(offset: str, label: str, schema: tuple | dict | None) -> str:
+    return f"{offset}{label}{_repr_schema(tuple(schema) if schema is not None else None)}\n"
+
+
+@functools.singledispatch
+def _repr_ir(ir: IR, *, offset: str = "") -> str:
+    return _repr_header(offset, type(ir).__name__.upper(), ir.schema)
+
+
+@_repr_ir.register
+def _(ir: GroupBy, *, offset: str = "") -> str:
+    keys = tuple(ne.name for ne in ir.keys)
+    return _repr_header(offset, f"GROUPBY {keys}", ir.schema)
+
+
+@_repr_ir.register
+def _(ir: Join, *, offset: str = "") -> str:
+    left_on = tuple(ne.name for ne in ir.left_on)
+    right_on = tuple(ne.name for ne in ir.right_on)
+    return _repr_header(offset, f"JOIN {ir.options[0]} {left_on} {right_on}", ir.schema)
+
+
+@_repr_ir.register
+def _(ir: Sort, *, offset: str = "") -> str:
+    by = tuple(ne.name for ne in ir.by)
+    return _repr_header(offset, f"SORT {by}", ir.schema)
+
+
+@_repr_ir.register
+def _(ir: Scan, *, offset: str = "") -> str:
+    first_path = ir.paths[0]
+    suffix = " ..." if len(ir.paths) > 1 else ""
+    label = f"SCAN {ir.typ.upper()} {first_path}{suffix}"
```

Comment:
Sorry I missed this earlier, but we do end up printing **many** `Scan` lines in the output when we are splitting by file. In order to trigger the `(repeated ... time)` logic, we can probably drop the `path` name in a follow-up PR. It seems like the schema information **should** be enough to indicate what dataset we are reading from (I hope?).

Reply:
Maybe we can just include the name of the first path? See https://github.com/rapidsai/cudf/pull/18752